### ISSUE TITLE
Segmentation fault compiling openssl-3.0.0-alpha1 on RHEL 6.10

### ISSUE THREAD

**Issue by satya-kotla** [openssl/openssl#11694]
I am compiling openssl-3.0.0-alpha1 on RHEL 6.10 server and getting Segmentation fault below. Can some one suggest a solution for this issue.

./config --prefix=/app/digital/openssl shared
Operating system: x86_64-whatever-linux2
Configuring OpenSSL version 3.0.0-alpha1 for target linux-x86_64
Using os-specific seed configuration
./config: line 910: 19681 Segmentation fault      (core dumped) /usr/bin/env __CNF_CPPDEFINES='' __CNF_CPPINCLUDES='' __CNF_CPPFLAGS='' __CNF_CFLAGS='' __CNF_CXXFLAGS='' __CNF_LDFLAGS='' __CNF_LDLIBS='' /usr/bin/perl ./Configure linux-x86_64 '--prefix=/app/digital/openssl' 'shared'

**Comment by mattcaswell** [comment]
What perl version are you running?

**Comment by satya-kotla** [comment]
perl version is 5.10.1

**Comment by mattcaswell** [comment]
I don't have access to RHEL 6.10 but this seems to be caused by a bug in older versions of perl. I am seeing a very similar crash in an older perl on a different platform, so it might be the same thing or might not.

Please could you try out the following workaround procedure and see if it works for you?

From a *clean* source tree please edit the file `apps/build.info` to remove these lines:

````
  DEPEND[${OPENSSLSRC/.c/.o}]=progs.h
  GENERATE[progs.c]=progs.pl -C $(APPS_OPENSSL)
  GENERATE[progs.h]=progs.pl -H $(APPS_OPENSSL)
  # progs.pl tries to read all 'openssl' sources, including progs.c, so we make
  # sure things are generated in the correct order.
  DEPEND[progs.h]=progs.c
````
Then run `config`/`Configure` as you normally would.

Before running make and while standing in the top of the source tree, execute the following:

````
touch apps/progs.c
perl apps/progs.pl -C apps/openssl >apps/progs.c
perl apps/progs.pl -H apps/openssl >apps/progs.h
````
Now run `make` and `make test` as normal.

Please let me know if this works.

**Comment by levitte** [comment]
Interesting workaround.
@mattcaswell, have you tried `perl -d Configure ... ` ? I'm not sure if it catches a core dump, but if it does, it would be nice to see exactly where that happens...

**Comment by mattcaswell** [comment]
> @mattcaswell, have you tried perl -d Configure ... ? I'm not sure if it catches a core dump, but if it does, it would be nice to see exactly where that happens...

Yes....you may recall you told me to do that once before, and I told you the answer :-) I don't think it was very enlightening (the platform I encounter this on is Solaris x86)

**Comment by levitte** [comment]
My memory is short sometimes...

Damn.

I wonder if I can lay my hands on an old enough perl for my machine.  Judging from what's been said here, the issue at hand doesn't seem to be tied to a particular platform...

**Comment by satya-kotla** [comment]
I followed the instructions exactly as mentioned above and now got below errors during make.
Please suggest how to fix the below issue.

/usr/bin/perl "-I." "-Iproviders/common/der" -Mconfigdata -Moids_to_c "util/dofile.pl" "-oMakefile" providers/common/der/der_dsa.h.in > providers/common/include/prov/der_dsa.h
String found where operator expected at /app/digital/softwares/openssl/util/perl/OpenSSL/OID.pm line 346, near "subtest "Checking '$_'""
        (Do you need to predeclare subtest?)
String found where operator expected at /app/digital/softwares/openssl/util/perl/OpenSSL/OID.pm line 373, near "subtest "Checking '$_'""
        (Do you need to predeclare subtest?)
syntax error at /app/digital/softwares/openssl/util/perl/OpenSSL/OID.pm line 346, near "subtest "Checking '$_'""
syntax error at /app/digital/softwares/openssl/util/perl/OpenSSL/OID.pm line 367, near "},"
Global symbol "@bad_cases" requires explicit package name at /app/digital/softwares/openssl/util/perl/OpenSSL/OID.pm line 372.
syntax error at /app/digital/softwares/openssl/util/perl/OpenSSL/OID.pm line 373, near "subtest "Checking '$_'""
syntax error at /app/digital/softwares/openssl/util/perl/OpenSSL/OID.pm line 387, near "},"
Compilation failed in require at providers/common/der/oids_to_c.pm line 16.
BEGIN failed--compilation aborted at providers/common/der/oids_to_c.pm line 16.
Compilation failed in require.
BEGIN failed--compilation aborted.
make[1]: *** [providers/common/include/prov/der_dsa.h] Error 255
make[1]: Leaving directory `/app/digital/softwares/openssl'
make: *** [build_sw] Error 2

**Comment by mattcaswell** [comment]
> I followed the instructions exactly as mentioned above and now got below errors during make.
Please suggest how to fix the below issue.

Ok - that's good news!! That means the workaround seems to work for the seg fault issue.

@levitte - any clues on this other issue?

**Comment by levitte** [comment]
Does it help if you install `perl-core`?  See NOTES.PERL:

https://github.com/openssl/openssl/blob/90113096186e185cd07ee1c4c7267d0e68e67793/NOTES.PERL#L23-L24

**Comment by tomiii00** [comment]
I confirm the workaround allows Configure to complete successfully but then I get the same build failure as @satya-kotla described. Not sure why the "touch" is requested since the shell redirection creates apps/progs.c before invoking perl.

**Comment by tomiii00** [comment]
FWIW I have both perl and perl-core RPMs installed.

**Comment by mattcaswell** [comment]
>  Not sure why the "touch" is requested since the shell redirection creates apps/progs.c before invoking perl.

Because progs.pl expects it to be there and complains if its not.

**Comment by levitte** [comment]
> FWIW I have both perl and perl-core RPMs installed.

Hmmm, ok

**Comment by levitte** [comment]
Ouch!

I found this in the COMPATIBILITY section of the Test::More man-page:

```
       subtests
           Subtests were released in Test::More 0.94, which came with Perl
           5.12.0. Subtests did not implicitly call "done_testing()" until
           0.96; the first Perl with that fix was Perl 5.14.0 with 0.98.
```

That's *really* annoying.

**Comment by tomiii00** [comment]
> > Not sure why the "touch" is requested since the shell redirection creates apps/progs.c before invoking perl.
> 
> Because progs.pl expects it to be there and complains if its not.

Yes, but the ">" operator causes the shell to create it prior to invoking perl, right? How else would perl's stdout get the right file descriptor?

**Comment by levitte** [comment]
Er, what "touch" are you talking about, and how did apps/progs.c get involved in this?

**Comment by tomiii00** [comment]
Matt's workaround above. He asked to try:

touch apps/progs.c
perl apps/progs.pl -C apps/openssl >apps/progs.c
perl apps/progs.pl -H apps/openssl >apps/progs.h

**Comment by levitte** [comment]
Oh, ah, that.  Yeah, that touch is not necessary.

**Comment by mattcaswell** [comment]
Hmmmm. When I tried this myself I needed to do the touch. But then I was experimenting with a few things...so possibly the final formulation I came up with doesn't need it.

**Comment by levitte** [comment]
The trouble with OID.pm is fixed in #11704

**Comment by satya-kotla** [comment]
Thanks @levitte. I could see pull request https://github.com/openssl/openssl/pull/11704 is still approval pending to merge to master branch.

But in the meanwhile, I have updated "util/perl/OpenSSL/OID.pm" file to remove UNIT TESTING portion in my local RHEL 6.10 server and compilation completed successfully without any further issues.

openssl version
OpenSSL 3.0.0-alpha2-dev  (Library: OpenSSL 3.0.0-alpha2-dev )

**Comment by mattcaswell** [comment]
Proper fix for the seg fault issue in #11737
